### ISSUE TITLE
Waku Status message not according to spec

### ISSUE THREAD

**Issue by kdeme** [status-im/status-go#1929]
# Problem
I've noticed while interop testing status-go with Nimbus wakunode that the handshake fails. 
I see two problems:

1. The keys from the `StatusOptions` struct get character encoded. So resulting rlp contains 48, 49, 50, etc. instead of 0, 1, 2, etc. I think the ABNF part of the spec dictates the actual integer values (if you compare it with other parts).
(see spec: https://github.com/vacp2p/specs/blob/master/specs/waku/waku.md#abnf-specification vs https://github.com/status-im/status-go/blob/develop/waku/handshake.go#L22 )

```
Reading rlp data of packet: StatusOptions
list{
list{
byte 48
blob(8) [3F50624DD2F1A9FC]
}
list{
byte 50
""
}
list{
byte 51
""
}
list{
byte 52
list{
""
""
""
}
}
}
```

1. Status-go disconnects from Nimbus wakunode because of issue 1. I didn't look too much in the code as to why, but I can confirm that the handshake succeeds when I adjust Nimbus wakunode to start the keys from 48.

```
DEBUG[04-06|17:37:43.797] Adding p2p peer id=cc746081f8465015 conn=inbound addr=127.0.0.1:59768 peers=3 name="Nimbus waku node"
DEBUG[04-06|17:37:43.797] confirm peer added ID=cc746081f8465015
DEBUG[04-06|17:37:43.797] peer added event peer=cc746081f8465015843c5c751a25ea1115cb5f4798974d4134ff3f39590cc823 inbound=true
DEBUG[04-06|17:37:43.798] Removing p2p peer id=cc746081f8465015 conn=inbound addr=127.0.0.1:59768 peers=2 duration=669.82µs req=false err="p [cc746081f8465015843c5c751a25ea1115cb5f4798974d4134ff3f39590cc823]: failed to decode status packet: rlp: call of ListEnd not positioned at EOL"
DEBUG[04-06|17:37:43.798] confirm peer dropped ID=cc746081f8465015
```

Why is 2. a problem? This is perhaps not directly written down in the spec( https://github.com/vacp2p/specs/blob/master/specs/waku/waku.md#status), but the idea was that when new k:v pairs in `StatusOptions` get added and the client doesn't know about these, then these get ignored. This was done for forward compatibility. 

So IMO, even with this bug, I think the handshake should still pass (fyi, according to Nimbus side, the handshake succeeds).

I will do a PR to explicitly write this down in the specification.

# Acceptance Criteria
Fix according to specification.

**Comment by andremedeiros** [comment]
> I will do a PR to explicitly write this down in the specification.

👏 

Can you update this PR when the spec gets refined so we can implement the fix on our end?

**Comment by adambabik** [comment]
1. This is true, status-go's waku implementation uses a string as a key type. I think it could be more clear in the spec but also it was accidental as the first implementation was using string values like `pow-requirement-key` instead of `0`. I blindly kept the data type and just changed the values.
[This is a place](https://github.com/status-im/status-go/blob/develop/waku/handshake.go#L76) where we should parse key as an integer and accordingly change the maps' types etc. Probably defining a key type as would be helpful (`type statusOptionKey int64`).

2. I didn't check it either but I suspect [this line](https://github.com/status-im/status-go/blob/develop/waku/handshake.go#L133-L136) to be a problem. status-go expects keys to be strings. Nimbus sends ints and this likely returns an error. It will be fixed when status-go switches to use the integer type as a key type.

Great found @kdeme!

**Comment by corpetty** [comment]
@adambabik when is this change expected?
 > It will be fixed when status-go switches to use the integer type as a key type.

Also, what are the consequences of this change?  Does fleet need updating, will we have compatibility issues of status apps not updating?

**Comment by kdeme** [comment]
You can track this PR for the possible spec addition regarding point 2.: https://github.com/vacp2p/specs/pull/112

**Comment by adambabik** [comment]
@corpetty I will talk to @Samyoul and the Core Team but this should be fixed asap.

> Also, what are the consequences of this change? Does fleet need updating, will we have compatibility issues of status apps not updating?

If we already released the app with Waku enabled by default, we need to first update the code so that it is backward compatible (recognize options with int and string keys), update fleets and then update the app so it starts sending keys as ints. 

If we haven't released the app with Waku enabled, we can do it without backward compatibility and simply change the key type to integer.

**Comment by Samyoul** [comment]
I'm going to make a pull for this. I'll drop the link here once it is up.

**Comment by Samyoul** [comment]
PR here https://github.com/status-im/status-go/pull/1931 . I will discuss this with @adambabik as there are a number of changes that I need to ensure are fit for purpose.

**Comment by Samyoul** [comment]
@kdeme Could you provide some sample handshakes that have failed for you? I am going to add them as part of the tests.

There are a number of complications with receipt of RLP encodings of mixed types for the key. I'm going to add some additional changes to the code to handle whether the incoming data was originally encoded as a string or an int. This will ensure that re-encoding will maintain its origin format.

**Comment by kdeme** [comment]
As described in the issue, failure was caused by invalid key integers being passed. Well, not actually invalid according to spec, just different than what status-go expects (different from the struct).

Considering the specification, there are many parts you could add tests for (perhaps status-go already had some of these, I don't know), e.g.:

Valid cases:
1. Good processing of keys & values as in the spec (0, 1, 2 ...).
2. Check if unknown keys don't cause an error & get ignored (see spec change for clarification: https://github.com/vacp2p/specs/pull/112).
3. Check if missing keys don't cause an error (and values get defaulted in this case).
4. Check if different ordering of key:value pairs doesn't cause an error.

Invalid cases:
1. Check if known keys with invalid value types do return an error.
2. Check if invalid key types returns an error.
3. Check other invalid types to the StatusOptions (not a list, or key:value item not a list).

**Comment by corpetty** [comment]
@Samyoul how is this going? I was curious if this has an expected timeline.

**Comment by Samyoul** [comment]
@corpetty I'm having trouble with this issue. I need to run my problems by the team.

Let me detail the issues, and maybe we can work through them.

### Uint Key Scenarios

I've put together a set of basic scenarios dealing with potential key values. They are as follows:

```golang
uKeys := []uint{0, 1, 2, 49, 50, 256, 257, 1000, 6000}
```

[View this test code](https://github.com/status-im/status-go/pull/1931/commits/9c1565624bf2d904d64c7b18445e22363ee4adbc#diff-89033975f57db1cdc14ed0fad9748368R41)

```golang
func TestKeyTypes(t *testing.T) {
	uKeys := []uint{
		0, 1, 2, 49, 50, 256, 257, 1000, 6000,
	}

	for i, uKey := range uKeys {
		fmt.Printf("test %d, for key '%d'", i+1, uKey)

		encodeable := []interface{}{
			[]interface{}{uKey, true},
		}
		data, err := rlp.EncodeToBytes(encodeable)
		spew.Dump(data, err)

		var optsDecoded statusOptions
		err = rlp.DecodeBytes(data, &optsDecoded)
		spew.Dump(optsDecoded, err)

		println("\n----------------\n")
	}
}
```

It outputs the following:

```text
test 1, for key '0'([]uint8) (len=4 cap=4) {
 00000000  c3 c2 80 01                                       |....|
}
(interface {}) <nil>
(uint64) 0
(*errors.errorString)(0xc042062030)(EOF)
(waku.statusOptions) {
 PoWRequirement: (*uint64)(0xc0421a3a88)(1),
 BloomFilter: ([]uint8) <nil>,
 LightNodeEnabled: (*bool)(<nil>),
 ConfirmationsEnabled: (*bool)(<nil>),
 RateLimits: (*waku.RateLimits)(<nil>),
 TopicInterest: ([]waku.TopicType) <nil>,
 keyType: (waku.statusOptionKeyType) 2
}
(interface {}) <nil>

----------------

test 2, for key '1'([]uint8) (len=4 cap=4) {
 00000000  c3 c2 01 01                                       |....|
}
(interface {}) <nil>
(uint64) 1
(interface {}) <nil>
(waku.statusOptions) {
 PoWRequirement: (*uint64)(<nil>),
 BloomFilter: ([]uint8) (len=1 cap=1) {
  00000000  01                                                |.|
 },
 LightNodeEnabled: (*bool)(<nil>),
 ConfirmationsEnabled: (*bool)(<nil>),
 RateLimits: (*waku.RateLimits)(<nil>),
 TopicInterest: ([]waku.TopicType) <nil>,
 keyType: (waku.statusOptionKeyType) 2
}
(interface {}) <nil>

----------------

test 3, for key '2'([]uint8) (len=4 cap=4) {
 00000000  c3 c2 02 01                                       |....|
}
(interface {}) <nil>
(uint64) 2
(interface {}) <nil>
(waku.statusOptions) {
 PoWRequirement: (*uint64)(<nil>),
 BloomFilter: ([]uint8) <nil>,
 LightNodeEnabled: (*bool)(0xc0421a3dd8)(true),
 ConfirmationsEnabled: (*bool)(<nil>),
 RateLimits: (*waku.RateLimits)(<nil>),
 TopicInterest: ([]waku.TopicType) <nil>,
 keyType: (waku.statusOptionKeyType) 2
}
(interface {}) <nil>

----------------

test 4, for key '49'([]uint8) (len=4 cap=4) {
 00000000  c3 c2 31 01                                       |..1.|
}
(interface {}) <nil>
(waku.statusOptions) {
 PoWRequirement: (*uint64)(<nil>),
 BloomFilter: ([]uint8) (len=1 cap=1) {
  00000000  01                                                |.|
 },
 LightNodeEnabled: (*bool)(<nil>),
 ConfirmationsEnabled: (*bool)(<nil>),
 RateLimits: (*waku.RateLimits)(<nil>),
 TopicInterest: ([]waku.TopicType) <nil>,
 keyType: (waku.statusOptionKeyType) 1
}
(interface {}) <nil>

----------------

test 5, for key '50'([]uint8) (len=4 cap=4) {
 00000000  c3 c2 32 01                                       |..2.|
}
(interface {}) <nil>
(waku.statusOptions) {
 PoWRequirement: (*uint64)(<nil>),
 BloomFilter: ([]uint8) <nil>,
 LightNodeEnabled: (*bool)(0xc0421dc0dc)(true),
 ConfirmationsEnabled: (*bool)(<nil>),
 RateLimits: (*waku.RateLimits)(<nil>),
 TopicInterest: ([]waku.TopicType) <nil>,
 keyType: (waku.statusOptionKeyType) 1
}
(interface {}) <nil>

----------------

test 6, for key '256'([]uint8) (len=6 cap=6) {
 00000000  c5 c4 82 01 00 01                                 |......|
}
(interface {}) <nil>
(uint64) 1
(interface {}) <nil>
(waku.statusOptions) {
 PoWRequirement: (*uint64)(<nil>),
 BloomFilter: ([]uint8) (len=1 cap=1) {
  00000000  01                                                |.|
 },
 LightNodeEnabled: (*bool)(<nil>),
 ConfirmationsEnabled: (*bool)(<nil>),
 RateLimits: (*waku.RateLimits)(<nil>),
 TopicInterest: ([]waku.TopicType) <nil>,
 keyType: (waku.statusOptionKeyType) 2
}
(interface {}) <nil>

----------------

test 7, for key '257'([]uint8) (len=6 cap=6) {
 00000000  c5 c4 82 01 01 01                                 |......|
}
(interface {}) <nil>
(uint64) 1
(interface {}) <nil>
(waku.statusOptions) {
 PoWRequirement: (*uint64)(<nil>),
 BloomFilter: ([]uint8) (len=1 cap=1) {
  00000000  01                                                |.|
 },
 LightNodeEnabled: (*bool)(<nil>),
 ConfirmationsEnabled: (*bool)(<nil>),
 RateLimits: (*waku.RateLimits)(<nil>),
 TopicInterest: ([]waku.TopicType) <nil>,
 keyType: (waku.statusOptionKeyType) 2
}
(interface {}) <nil>

----------------

test 8, for key '1000'([]uint8) (len=6 cap=6) {
 00000000  c5 c4 82 03 e8 01                                 |......|
}
(interface {}) <nil>
(uint64) 3
(interface {}) <nil>
(waku.statusOptions) {
 PoWRequirement: (*uint64)(<nil>),
 BloomFilter: ([]uint8) <nil>,
 LightNodeEnabled: (*bool)(<nil>),
 ConfirmationsEnabled: (*bool)(0xc0421dc5a0)(true),
 RateLimits: (*waku.RateLimits)(<nil>),
 TopicInterest: ([]waku.TopicType) <nil>,
 keyType: (waku.statusOptionKeyType) 2
}
(interface {}) <nil>

----------------

test 9, for key '6000'([]uint8) (len=6 cap=6) {
 00000000  c5 c4 82 17 70 01                                 |....p.|
}
(interface {}) <nil>
(uint64) 23
(interface {}) <nil>
(waku.statusOptions) {
 PoWRequirement: (*uint64)(<nil>),
 BloomFilter: ([]uint8) <nil>,
 LightNodeEnabled: (*bool)(<nil>),
 ConfirmationsEnabled: (*bool)(<nil>),
 RateLimits: (*waku.RateLimits)(<nil>),
 TopicInterest: ([]waku.TopicType) <nil>,
 keyType: (waku.statusOptionKeyType) 2
}
(interface {}) <nil>
```

#### Analysis

|Test number|Inputs|Comments|
|---|---|---|
|1|`uint(0)`|Output is correct, except it generates an EOF error that I ignore, this is due to a hack that I've implemented, to get indexes of 0 to work, but it is ugly. For some reason the application parses `uint(0)` as a null value. **I don't know why this is happening**|
|2|`uint(1)`|No problem|
|3|`uint(2)`|No problem|
|4|`uint(49)`|Problem key is interpreted as a string of value `"1"` . This is happening because a byte of value 49 has a string value of `1`. **I haven't been able to think of a solution for this**|
|5|`uint(50)`|Problem key is interpreted as a string of value `"2"` . This is happening because a byte of value 50 has a string value of `2`. **I haven't been able to think of a solution for this**|
|6|`uint(256)`|Problem key is interpreted as a uint of value `1`. This is happening because the first byte in the `[]byte` representing the key value is of value `1`. This is likely due to not using the correct parsing algo|
|7|`uint(257)`|Problem key is interpreted as a uint of value `1`. This is happening because the first byte in the `[]byte` representing the key value is of value `1`. This is likely due to not using the correct parsing algo|
|8|`uint(1000)`|Problem key is interpreted as a uint of value `3`. This is happening because the first byte in the `[]byte` representing the key value is of value `3`. This is likely due to not using the correct parsing algo|
|9|`uint(6000)`|Problem key is interpreted as a uint of value `23`. This is happening because the first byte in the `[]byte` representing the key value is of value `23`. This is likely due to not using the correct parsing algo|

### Preferring Strings

The main algo I've written to determine the key values is below. [See code here](https://github.com/status-im/status-go/pull/1931/commits/9c1565624bf2d904d64c7b18445e22363ee4adbc#diff-75951de319dc41e8224b6b9cca9148b8R207)

```golang
	// decode into bytes, detect if bytes can be parsed as a string and from a string to a uint
	var bKey []byte
	if err := s.Decode(&bKey); err != nil {
		return 0, 0, err
	}

	// Parse string into uint
	uKey, err := strconv.ParseUint(string(bKey), 10, 64)
	if err == nil {
		return statusOptionKey(uKey), sOKTS, err
	}

	// If statusOptionKey (uint) can be decoded return it
	buf := bytes.NewBuffer(bKey)
	uintKey, c := binary.ReadUvarint(buf)
	spew.Dump(uintKey, c)

	return statusOptionKey(uintKey), sOKTU, nil
```

You will notice that I prioritise parsing the key's byte value into string, this is because byte values for strings start at 48 meaning we get at least 48 keys before the parsing starts to mess up.

If I had implemented the algo preferring uints then no string type would be detected because a byte can always be interpreted as a uint.

**Example:**
`string("0")`, encodes to `byte(48)`, decodes to `uint(48)`.
`string("111")`, encodes to `[]byte(131, 49, 49, 49)`, decode to `uint(3223857)`.
This means an expected index of 0 will be returned as 48. An expected index of 111 will be returned as 3223857

### My Problems

- For some reason the application parses a key value of `uint(0)` as a null value. **I don't know why this is happening**
- Keys with a byte of value 47 between 58 have a string value of `0-9`, there is no way to determine whether the key should be a uint of 48 or a string of `"0"`. **I haven't been able to think of a backwards compatible solution for this**
- Need to find the correct algo for parsing uints of value greater than 127, not a major issue.

### Summary

These problems could be the result of me missing something and being slow, so I am happy to be corrected or pointed to the thing. I will walk that walk of shame 😄 .

On a plus note all string type keys now parse as expected, both in decoding and re-encoding.

**Comment by Samyoul** [comment]
@corpetty and @kdeme https://github.com/status-im/status-go/pull/1938 is replacing my previous branch with a different approach to the problem. Also see https://github.com/vacp2p/specs/pull/113 .

There is some init functionality in my old branch that I'd like to keep as I think it is a better implementation of variable initialisation. But the rest is redundant.

Basically we're implementing distinct versions for Waku. We're retrospectively changing the specs for `waku/0`, and implementing a `waku/1` that implements uint encoded RLP keys.

**Comment by Samyoul** [comment]
@kdeme , @corpetty I believe this issue is now resolved with the merging of @cammellos' PR https://github.com/status-im/status-go/pull/1956. Can you check and let me know if you the issue appears resolved?

**Comment by kdeme** [comment]
@Samyoul Yes, it looks OK now. Can be closed.

**Comment by Samyoul** [comment]
Resolved with #1956
